### ISSUE TITLE
delay line modulation in fdn reverb node causes distorsion and other od audio effects

### ISSUE THREAD

**Issue by roelvdwal** [camlorn/libaudioverse#23]
when using delay_modulation_depth with the fdn reverb node, the audio gets distorted and starts clicking. For a recording of this, download
https://dl.dropboxusercontent.com/u/34600650/reverb_test.flac
The code I pasted at the python console is in the attachment.
[entered_code.txt](https://github.com/camlorn/libaudioverse/files/274889/entered_code.txt)

**Comment by camlorn** [comment]
This is a more general problem.  You can also see it if you try to modulate any of the existing delay lines by more than a bit.

The solution is to set them to very, very small numbers or to not use them.  Generally, you can make it good enough with just `t60` and the filter cutoff properties.

I'm going to fix this but probably post-0.9, as it requires writing some stuff that's actually kind of deep mathematics.

How important are they to you?  Given that I can't fix these quickly or soon, I should probably remove them for 0.9.  The entire code there needs rewriting.  If you could do some further experiments and determine how much you want them, that would be very helpful.

Also, on a sidenote, you might find the recorder node helpful.  You can configure it to send any audio passing through it to a wave file.  Not needed here, obviously, but it could be helpful as your project gets more complicated: adding a keystroke to record your game's audio or whatever to out.wav in the current directory can be done in one function call.  I should highlight this more prominently in the manual.

**Comment by roelvdwal** [comment]
O, I thought I was the only one with this issue. That's why I posted it. I've never used delay modulation before, so I don't know if I'll need it. I'll also make sure to use the recorder node next time.

**Comment by camlorn** [comment]
Yeah.  There's stuff that's known that isn't on the GitHub issue tracker.  I hit most of the major ones without bothering to report against myself, so the issues I open here are the "we need to do this...eventually" issues.  Obviously I don't treat issues from others this way.  You're the only one having this one because you're the only one besides me who has ever played with those knobs to my knowledge, but it is known and will be fixed.

Reverbs are an art, not a science.  In future, there will actually probably be other designs.  I don't know how important the delay line modulation is either, but it was an idea mentioned in the place that gave me the background math that eventually became what we have, so I threw it in.  Then it didn't work nearly as well as I would expect, I went "I can fix this but it needs...", and never bothered to remove it because it's fixable.  Eventually.

I don't care how you make the demos; I was merely mentioning the recorder node on the off chance that it's helpful to you in the future.  You can't always make a demo from the REPL, but there's no right or wrong way to do it as long as it works.

If this hadn't been an issue I already knew about, you gave me exactly what I needed to solve the problem and in a convenient format.  Usually getting all the info you gave (even a tenth of the info you gave) requires me having a long conversation.  If I have given the impression that I'm dissatisfied with your original report, that is most certainly not the case.  Quite the opposite, honestly.  This is a very good bug report.

**Comment by roelvdwal** [comment]
O, I didn't have the impression that you were dissatisfied with the repport at all. Now looking back on my previous comment I see why you thought I did. Also, I red that fft reverb is also possible, but I've not been able to figure out how it works. Is it worth it to invest my time in getting fft reverb or are the gains so little that it isn't worth spending my time on?

**Comment by camlorn** [comment]
If the following is patronizing, I apologize; I operate under the assumption that people don't have DSP/audio effects backgrounds until proven otherwise.

Fft reverb is possible, but you currently need to write it yourself using an `FftConvolverNode`.  It works, but I haven't used it in a very long time and should have but didn't write an example.

Now, that said, in comparison to the FdnReverb, it is and will always be incredibly slow.  It's a`O(n log n)` algorithm, so it just eats up more CPU the longer your reverb gets.  By comparison, `FdnReverb` is a constant-time algorithm.  Even if you set `t60` to ridiculous values like 200 seconds, it's still going to use the same CPU.

As we're getting kind of off topic, I'd like to go ahead and point you at the [Google group](https://groups.google.com/a/camlorn.net/d/forum/libaudioverse).  If you make a thread there then I'll be happy to help you work out the code for FFT reverb.
